### ISSUE TITLE
Internal Candle Error at Backprop: 'called `Option::unwrap()` on a `None` value'

### ISSUE THREAD

**Issue by LunaticWyrm467** [huggingface/candle#863]
### The Problem
Candle crashes with an internal unhandled error while attempting to backpropagate through my module:
```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', C:\Users\[USER]\.cargo\git\checkouts\candle-0c2b4fa9e5801351\e82fcf1\candle-core\src\backprop.rs:137:43
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library\std\src\panicking.rs:578
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library\core\src\panicking.rs:67
   2: core::panicking::panic
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library\core\src\panicking.rs:117
   3: enum2$<core::option::Option<candle_core::tensor::Tensor> >::unwrap<candle_core::tensor::Tensor>
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca\library\core\src\option.rs:950
   4: candle_core::tensor::Tensor::backward
             at C:\Users\[USER]\.cargo\git\checkouts\candle-0c2b4fa9e5801351\e82fcf1\candle-core\src\backprop.rs:137
   5: candle_nn::optim::Optimizer::backward_step<candle_nn::optim::AdamW>
             at C:\Users\[USER]\.cargo\git\checkouts\candle-0c2b4fa9e5801351\e82fcf1\candle-nn\src\optim.rs:21
```

### MRP
Unfortunately, I tried to separate the training() functions from the module itself, but that did not yield the error, so I had to include the full Key-Value memory network I was working on, along with a few more components such as some traits and math helper functions. Sorry if this is a bit long, but it is contained in a single main.rs file, along with the .toml rust config file.
[MRP.zip](https://github.com/huggingface/candle/files/12620510/MRP.zip)

### MRP Context
I've been working on a specialized Key-Value Memory Network for a deep learning agent of mine, which saves embedded key vectors into a key memory matrix of size (MEM_SIZE, EMBED_SIZE), along with their associated values in a matrix of size (MEM_SIZE, VAL_SIZE). Memories are retrieved via finding the closest matching embedded key vector to the embedded query vector using cosine similarity, and then fetching the key's associated value.
The self-supervised training function, from where the crash is spawned via the backprop method, is simply calculating the correlation of the key's associated values, attempting to find the least/most correlated data in memory, and shapes the MemoryCircuits (the embedding network) to produce meaningful vectors that are clustered around associative data, and are farther away from unrelated data. The FusionCircuits simply learn off of the difference between key vectors generated before and after the loss was applied, and attempt to fit the old keys to represent their data better.

### Expected Behaviour
If I am doing something wrong on my end, it should warrant a proper crash message, or should be handled similarly to other errors that are propagated throughout the candle crate. However, I can't really tell why this would be an issue, as I am simply backpropagating an error value through the network.

**Comment by LaurentMazare** [comment]
Thanks for reporting this, it seems that this was cause by the backprop code having improper dependencies specified for argmax/argmin (as the code here was shared with max/min). This should hopefully be fixed in the latest github version through #865 .

**Comment by LunaticWyrm467** [comment]
Thanks! I'll do some more testing once Candle updates.

**Comment by LaurentMazare** [comment]
Closing this as hopefully fixed, feel free to re-open if you run into further issues. I'm releasing v0.2.2 of the crate right now which should include the bugfix.

**Comment by LunaticWyrm467** [comment]
Gotcha, thanks so much - I'll be sure to report future issues.
